[Bryan Ong] iP

likeabowx opened the pull request:
# Puke - The Chat Bot
> “Unless You Puke, Faint or Die, Keep Going!” — Jillian Michaels ([Source](https://www.goodreads.com/quotes/110687-unless-you-puke-faint-or-die-keep-going))

Well, why not puke AND keep going? :vomiting_face::running_man:
Puke keeps you going by helping you manage **_all_** your tasks in one place! 

Puke is...
- text-based, no need for clicking
- super easy to learn

Get your life organised in ~~minutes~~ _seconds_ for **FREE**!

Let's get going now, all you need to do is,

1. download it from [here](https://github.com/likeabowx/ip/releases/download/A-Jar/main.jar)
2. double-click it
3. start adding your tasks to Puke
4. get up and start clearing the list! :wink:

Awesome features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
- [ ] Mind reading (probably not so soon but it's a possibility)
---
If you are into Java, you can use it to practice Java too! Here's the ``` main``` method
```java
public class PukeMain {
    public static void main(String[] args) {
        new Puke("data/puke.txt").run();
    }
}
```

Commit:
Add Gradle support

Commit:
Implemented level 1 - Greet, Echo, Exit

Commit:
Renamed file to Puke

Commit:
Implemented level 2 - Add, List

Commit:
Implemented level 3 - Mark as Done;

Commit:
Implemented level 3 - ToDos, Events, Deadlines

Commit:
Implemented level 4 - ToDos, Events, Deadlines

Commit:
Level 4 - Implemented inheritance

Commit:
Created files for I/O redirection testing

Commit:
Implemented level 5 - Handle errors

Commit:
Implemented level 6 - Delete

Commit:
Included test for error handling and deleting task

Commit:
Perform minor changes to comply Java Coding Standard

Commit:
Change method names to camelCase

Commit:
Implement level 7 - Save

Commit:
Implement level 8 - Dates and Times

Commit:
Merge branch 'branch-Level-8'

Commit:
Rename 'DukeException' to 'PukeException'

Commit:
Extract 4 new classes - Ui, Parser, Storage, TaskList

Commit:
Divide classes into packages

Commit:
Remove unused ui object from processInput method

Commit:
Add JUnit tests

Commit:
Add MANIFEST.MF

Commit:
Implement level 9 - Find

Commit:
Ensure the code comply with coding standard

Commit:
Change indent to 4 spaces

Commit:
Add JavaDoc comments

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/puke/task/Task.java
#	src/main/java/puke/task/TaskList.java

Commit:
Merge branch 'add-gradle-support' into Branch-A-Gradle

jiewei98 commented on src/main/java/puke/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return String representation of the task for the storage file.
+     */
+    abstract String toSaveString();
```
Would this be clearer if you named it toStringStorage()?

jiewei98 commented on src/main/java/puke/task/TaskList.java:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @return String representation of all tasks to save.
+     */
+    public String generateStorageData() {
```
an alternative name to this method could be convertTasksToString()?

jiewei98 commented on src/main/java/puke/ui/Ui.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Prints the exit message.
+     */
+    public void printExit() {
```
Since you have a printWelcomeMessage() this could be printGoodbyeMessage()?

jiewei98 commented on src/main/java/puke/task/TaskList.java:
```diff
@@ -0,0 +1,229 @@
+     * @return String representation of the task list, or a message if the list is empty.
+     * @throws PukeException If the list of task is empty.
+     */
+    public String listTasks() throws PukeException {
```
displayTasks() could be good as listTasks could be confused with task list?

jiewei98 reviewed (commented):
LGTM. The code is well formatted and very neat.

Commit:
Automate project builds using Gradle

Commit:
Merge branch 'Branch-A-Gradle'

Commit:
Fix build.gradle file

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Rectify coding style violations using checkStyle

Commit:
Merge branch 'branch-A-CheckStyle'

alexteo98 commented on src/main/java/puke/task/TaskList.java:
```diff
@@ -0,0 +1,229 @@
+     * @throws PukeException If the task number is invalid, or the task is already marked as done.
+     */
+    public String markTask(int taskNo) throws PukeException {
+        if (!this.isValidTask(taskNo)) {
+            throw new PukeException(String.format("%d is not a valid task number!", taskNo));
+        }
+
+        Task t = tasks.get(taskNo - 1);
+        if (t.isDone()) {
+            throw new PukeException(String.format("Task #%d is already marked as done!", taskNo));
+        }
```
Good use of guard clauses!

Commit:
Add a GUI to Duke

Commit:
Merge branch 'branch-Level-10'

alexteo98 commented on src/main/java/puke/task/Event.java:
```diff
@@ -0,0 +1,43 @@
+        this.at = datetime;
+    }
+
+    /**
+     * Generates a string to save the task information on the storage file.
+     *
+     * @return String representation of the task for the storage file.
+     */
+    public String toSaveString() {
+        return "E@@" + (this.isDone() ? "1@@" : "0@@")
+                + this.name + "@@" + this.at.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+    }
+
+    /**
+     * Returns a string representation of the task.
+     *
+     * @return String representation of the task.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString()
+                + String.format(" (at: %s)", this.at.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")));
+    }
+}
```
Perhaps you can move these functions to the abstract class Task.java to reduce code duplication.

alexteo98 commented on src/main/java/puke/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+ */
+public abstract class Task {
+    protected final String name;
+    protected int status; // 0: undone, 1: done
```
Would it be better if it is a boolean type?

alexteo98 commented on src/main/java/puke/parser/Parser.java:
```diff
@@ -0,0 +1,60 @@
+            argument = inputs[1]; // remaining input (if any)
+        }
+
+        switch (command) {
+        case "bye":
+            return null;
+        case "list":
+            if (argument == null) {
+                return tasks.listTasks();
+            } else {
+                throw new PukeException("I don't need any argument for list..");
+            }
+        case "mark":
+        case "unmark":
+            try {
+                int taskNo = Integer.parseInt(argument);
+                if (command.equals("mark")) {
+                    return tasks.markTask(taskNo);
+                } else {
+                    return tasks.unmarkTask(taskNo);
+                }
+            } catch (NumberFormatException e) {
+                throw new PukeException("I'll need a valid task number for it..");
+            }
+        case "todo":
+        case "deadline":
+        case "event":
+            return tasks.addTask(command, argument);
+        case "delete":
+            return tasks.deleteTask(Integer.parseInt(argument));
+        case "find":
+            return tasks.findTasks(argument);
+        default:
+            throw new PukeException("Are you sure you're making sense?");
+        }
```
Well done on doing the correct indentation for this switch block!

alexteo98 reviewed (commented):
Well done overall! Variable naming was very clear and was easy to understand 👍

Commit:
Update main class name in build.gradle

Commit:
Use assertions

likeabowx committed:
Merge pull request #1 from likeabowx/branch-A-Assertions

Add assertions

Commit:
Merge branch 'master' into branch-A-Assertions

Commit:
Improve code quality

There were code of low quality that violated certain topics covered in
the module content.

Improvements made includes using SLAP to extract methods, using guard
clauses to make the happy path more prominent, and improving comments
made in the code, and removing unsafe practices.

These improvements will help increase the readability of the code.

Commit:
Use streams for searching task

It is inefficient to loop through the array to search for tasks with
matching keywords.

Make use of Java streams to improve the searching for keyword in the
list of tasks.

likeabowx committed:
Merge pull request #2 from likeabowx/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' into branch-A-CodeQuality

likeabowx committed:
Merge pull request #3 from likeabowx/branch-A-Streams

Use streams for searching task

Commit:
Merge branch 'master' into branch-A-Streams

Commit:
Update the GUI

Label is used as the control to display text.

The use of label limits each message to 7 lines, and will become ...
after the message exceeds 7 lines (for example when listing more than
6 tasks).

Text is used instead to display the messages.

Text will dynamically resize the height when the message gets longer.

In addition, the photos of Puke and the user are updated.

Commit:
Implement sort command to sort by date, name, and type

Commit:
Implement help command

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Merge branch 'branch-A-Streams'

Commit:
Improve the GUI

Commit:
Add title to the GUI window

Commit:
Add user guide and UI screenshot

likeabowx committed:
Set theme jekyll-theme-tactile

likeabowx committed:
Set theme jekyll-theme-minimal

likeabowx committed:
Set theme jekyll-theme-minimal

likeabowx committed:
Set theme jekyll-theme-tactile

Commit:
Update README.md

Commit:
Merge branch 'master' of https://github.com/likeabowx/ip

likeabowx committed:
Update README.md

likeabowx committed:
Update README.md

likeabowx committed:
Update README.md
